build-tools/packages/build-cli/src/commands/modify/fluid-imports.ts

```diff
@@ -0,0 +1,334 @@
+	// Instead, we save files individually.
+	await Promise.all(fileSavePromises);
+}
+
+/**
+ * Parses an import declaration into its module specifier and subpath.
+ *
+ * @param importDeclaration - the import declaration to check.
+ * @returns a tuple of `[module specifier, subpath]`
+ */
+function parseImport(importDeclaration: ImportDeclaration): [string, string] {
+	const moduleSpecifier = importDeclaration.getModuleSpecifierValue();
+	// log?.verbose(`Found a Fluid import: '${moduleSpecifier}'`);
+	const modulePieces = moduleSpecifier.split("/");
+	const moduleName = modulePieces.slice(0, 2).join("/");
+	const subpath = modulePieces.length === 3 ? modulePieces[2] : "public";
+	// log?.verbose(`subpath: ${subpath}`);
+	return [moduleName, subpath];
+}
+
+function isFluidImport(importDeclaration: ImportDeclaration): boolean {
+	const moduleSpecifier = importDeclaration.getModuleSpecifierValue();
+	return (
+		moduleSpecifier.startsWith("@fluid") ||
+		["fluid-framework", "tinylicious"].includes(moduleSpecifier)
+	);
+}
+
+interface MemberDataRaw {
+	level: ApiLevel;
+	kind: string;
+	name: string;
+}
+
+type MemberData = Partial<Record<ApiLevel, string[]>>;
+type MapData = Map<string, MemberData>;
+
+function isKnownLevel(level: string): level is ApiLevel {
+	return (knownLevels as readonly string[]).includes(level);
+}
+
+function ensureLevel(entry: MemberData, level: keyof MemberData): string[] {
+	const entryData = entry[level];
+	if (entryData !== undefined) {
+		return entryData;
+	}
+	const newData: string[] = [];
+	entry[level] = newData;
+	return newData;
+}
+
+async function loadData(dataFile: string): Promise<MapData> {
+	// Load the raw data file
+	// eslint-disable-next-line unicorn/no-await-expression-member
+	const rawData: string = (await readFile(dataFile)).toString();
+	const apiLevelDataRaw: Record<string, MemberDataRaw[]> = JSON5.parse(rawData);
+
+	// Transform the raw data into a more useable form
+	const apiLevelData = new Map<string, MemberData>();
+	for (const [moduleName, members] of Object.entries(apiLevelDataRaw)) {
+		const entry = apiLevelData.get(moduleName) ?? {};
+		for (const member of members) {
+			const { level } = member;
+			if (!isKnownLevel(level)) {
+				throw new Error(`Unknown API level: ${level}`);
+			}
+			ensureLevel(entry, level).push(member.name);
+		}
+		apiLevelData.set(moduleName, entry);
+	}
+	return apiLevelData;
+}
+
+/**
+ * Delete any header comments at the beginning of the file. Return the removed text.
+ */
+function removeFileHeaderComment(sourceFile: SourceFile): string {
+	const firstNode = sourceFile.getChildAtIndex(0);
+	const headerComments = firstNode.getLeadingCommentRanges();
+	// const headerComments = imports[0].getLeadingCommentRanges();
+	const headerText = `${headerComments.map((comment) => comment.getText()).join("\n\n")}\n\n`;
+	const [start, end] = [firstNode.getPos(), firstNode.getEnd()];
+	sourceFile.replaceText([start, end], sourceFile.getChildAtIndex(0).getText());
```

Comment:
Does seem like there are probably cases where comments related to the import are left at start and the import line moves. But I don't really know the structure. (LeadingCommentRanges being Ranges raised a flag.) If this is really a case, the comments would also be nice.

Reply:
Yes, I think this is a fact of life unfortunately. This approach will at least ensure that all leading comment text is still there after rewrite, so fixing up anything that wasn't moved should be straightforward. We may come across cases I haven't considered, though.